plugins/ecs-serviceconnect/e2eTests/e2e_test.go

```diff
@@ -0,0 +1,288 @@
+
+	// Create a network namespace to mimic the container's network namespace.
+	targetNS, err := netns.NewNetNS(nsName)
+	require.NoError(t, err,
+		"Unable to create the network namespace that represents the network namespace of the container")
+	defer targetNS.Close()
+
+	// Construct args to invoke the CNI plugin with.
+	execInvokeArgs := &invoke.Args{
+		ContainerID: containerID,
+		NetNS:       targetNS.GetPath(),
+		IfName:      ifName,
+		Path:        os.Getenv("CNI_PATH"),
+	}
+
+	// Execute the "ADD" command for the plugin.
+	execInvokeArgs.Command = "ADD"
+	err = invoke.ExecPluginWithoutResult(
+		pluginPath,
+		netConfData,
+		execInvokeArgs)
+	require.NoError(t, err, "Unable to execute ADD command for ecs-serviceconnect cni plugin")
+
+	targetNS.Run(func() error {
+		// Validate IP rules successfully added.
+		validateIPRulesAdded(t, netConf)
+		return nil
+	})
+
+	// Execute the "DEL" command for the plugin.
+	execInvokeArgs.Command = "DEL"
+	err = invoke.ExecPluginWithoutResult(
+		pluginPath,
+		netConfData,
+		execInvokeArgs)
+	require.NoError(t, err, "Unable to execute DEL command for ecs-serviceconnect cni plugin")
+
+	targetNS.Run(func() error {
+		// Validate IP rules successfully deleted.
+		validateIPRulesDeleted(t, netConf)
+		return nil
+	})
+}
+
+// validateIPRulesAdded validates IP rules created in the target network namespace.
+func validateIPRulesAdded(t *testing.T, netConf *config.NetConfig) {
+	for proto, mustExist := range getProto(netConf) {
+		iptable, err := iptables.NewWithProtocol(proto)
+		require.NoError(t, err, "Unable to initialize iptable")
+
+		validateIngressIptableRulesAdded(t, iptable, netConf,
+			mustExist && len(netConf.IngressRedirectToListenerPortMap) > 0)
+
+		//TODO: Add support for BRIDGE_TPROXY
+		switch netConf.Mode {
+		case config.AWSVPC, config.BRIDGE_DNAT:
+			validateEgressIptableRulesAdded(t, iptable, proto, netConf,
+				mustExist && netConf.EgressPort != 0)
+		}
+	}
+}
+
+// validateIngressIptableRulesAdded validates the IP table rules that are added for handling ingress traffic
+func validateIngressIptableRulesAdded(t *testing.T, iptable *iptables.IPTables, netConf *config.NetConfig, mustExist bool) {
+	for redirectPort, listenerPorts := range netConf.IngressRedirectToListenerPortMap {
+		exist, _ := iptable.Exists("nat", ingressChain, "-p", "tcp", "-m", "multiport", "--dports", listenerPorts,
+			"-j", "REDIRECT", "--to-port", redirectPort)
+		require.Equal(t, mustExist, exist, "Mismatch of expected rules to redirect ingress ports")
+
+		exist, _ = iptable.Exists("nat", "PREROUTING", "-p", "tcp", "-m", "addrtype", "!",
+			"--src-type", "LOCAL", "-j", ingressChain)
+		require.Equal(t, mustExist, exist, "Mismatch of expected rules to redirect non-local traffic")
+	}
+}
+
+// getCidr returns the Cidr for the protocol
+func getCidr(proto iptables.Protocol, netConf *config.NetConfig) string {
+	var cidr string
+	if proto == iptables.ProtocolIPv4 {
```

Comment:
do we support dual stack? or it's always either or?

Reply:
Currently, I have assumed that IPV4 would always be present and IPV6 address is optional since we dont support V6-only traffic yet.